pkg/controller/manager/sync_action.go

```diff
@@ -86,229 +89,195 @@ func runSync(ctx context.Context, cluster *scyllav1.ScyllaCluster, authToken str
 
-	backupActions, err := syncBackupTasks(clusterID, cluster, syncer, state)
+	backupActions, err := syncBackupTasks(clusterID, cluster, state)
 	if err != nil {
-		return nil, errors.Wrap(err, "runSync repair tasks")
+		errs = append(errs, fmt.Errorf("can't sync backup tasks: %w", err))
 	}
-
 	actions = append(actions, backupActions...)
 
+	err = utilerrors.NewAggregate(errs)
+	if err != nil {
+		return nil, err
+	}
+
 	return actions, nil
 }
 
-func syncBackupTasks(clusterID string, cluster *scyllav1.ScyllaCluster, syncer stateCache, managerState *state) ([]action, error) {
+func syncBackupTasks(clusterID string, cluster *scyllav1.ScyllaCluster, managerState *state) ([]action, error) {
+	var errs []error
 	var actions []action
 
 	for _, bt := range cluster.Spec.Backups {
-		btCopy := *bt.DeepCopy()
-		backupTask := &BackupTask{BackupTaskSpec: btCopy}
+		action, err := syncBackupTask(clusterID, managerState, BackupTaskSpec(*bt.DeepCopy()))
+		if err != nil {
+			errs = append(errs, fmt.Errorf("can't sync backup task %q: %w", bt.Name, err))
+			continue
+		}
 
-		for _, managerTask := range managerState.BackupTasks {
-			if syncer.taskID(backupTask.Name) == managerTask.ID {
-				evaluateDates(backupTask, managerTask)
-			}
+		if action != nil {
+			actions = append(actions, action)
 		}
+	}
 
-		if syncer.shouldCreateTask(backupTask.Name) {
-			mt, err := backupTask.ToManager()
-			if err != nil {
-				return nil, errors.Wrap(err, "transform to manager task")
-			}
-			actions = append(actions, &addTaskAction{
-				clusterID: clusterID,
-				task:      mt,
-				taskSpec:  btCopy,
-			})
-		} else if syncer.shouldUpdateTask(backupTask.Name) {
-			backupTask.ID = syncer.taskID(backupTask.Name)
-
-			update := false
-			for _, managerTask := range managerState.BackupTasks {
-				if managerTask.ID == backupTask.ID {
-					update = !reflect.DeepEqual(backupTask, managerTask)
-					break
-				}
-			}
-			if update {
-				mt, err := backupTask.ToManager()
-				if err != nil {
-					return nil, errors.Wrap(err, "transform to manager task")
-				}
+	err := utilerrors.NewAggregate(errs)
+	if err != nil {
+		return nil, err
+	}
 
-				actions = append(actions, &updateTaskAction{
-					clusterID: clusterID,
-					task:      mt,
-					taskSpec:  btCopy,
-				})
-			}
+	return actions, nil
+}
+
+func syncBackupTask(clusterID string, managerState *state, backupTask BackupTaskSpec) (action, error) {
+	managerTaskStatus, ok := managerState.BackupTasks[backupTask.Name]
+	if !ok {
+		managerClientTask, err := backupTask.ToManager()
+		if err != nil {
+			return nil, fmt.Errorf("can't transform to manager task: %w", err)
 		}
+
+		return &addTaskAction{
+			clusterID: clusterID,
+			task:      managerClientTask,
+		}, nil
 	}
 
-	return actions, nil
+	if managerTaskStatus.ID == nil {
+		// Sanity check.
+		return nil, fmt.Errorf("manager task status is missing an id")
+	}
+
+	evaluateDates(&backupTask, &managerTaskStatus)
```

Comment:
I think I don't understand the deepcopy part - we want the change introduced by `evaluateDates` to be propagated both to the comparison and to the update action, otherwise we'd overwrite the start date.
I isolated status comparison to a separate func.

Reply:
right, added deepcopy